File: nshm/src/tests/test_grid.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "nshm_api.h"
#include "nshm_grid.h"

void make_small_grid(NSHM_Grid *_grid);

int main(int argc, char ** argv) {
	NSHM_Grid grid;
	int status;

	char * query_delete_meta = "\nDELETE FROM %s WHERE %s = %d;\n";
	char * query_delete_data = "DELETE FROM %s WHERE %s = %d;\n\n";

	if ( argc < 4 ) {
		printf("Usage: %s <user> <pass> <tnsalias>\n", argv[0]);
		return EXIT_FAILURE;
	}

	nshm_initialize(argv[1], argv[2], argv[3]);

	printf("#############################################################\n");
	printf("## Fetching an Agrid                                       ##\n");
	printf("#############################################################\n");
	status = nshm_get_agrid(&grid, "adapt_cy");
	nshm_print_grid(&grid);
	nshm_free_grid(&grid);

	printf("#############################################################\n");
	printf("## Fetching a Bgrid                                        ##\n");
	printf("#############################################################\n");
	status = nshm_get_bgrid(&grid, "BVAL_.1");
	nshm_print_grid(&grid);
	nshm_free_grid(&grid);

	printf("#############################################################\n");
	printf("## Fetching a Mmax                                         ##\n");
	printf("#############################################################\n");
	status = nshm_get_mmax(&grid, "gm_j_6p6_7p1");
	nshm_print_grid(&grid);
	nshm_free_grid(&grid);

	printf("#############################################################\n");
	printf("## Writing small sample grids.                             ##\n");
	printf("#############################################################\n");

	make_small_grid(&grid);

	printf("Inserting an agrid...");
	if (nshm_put_agrid(&grid)==NSHM_RETURN_SUCCESS) {
		printf("success!\n"); } else { printf("failed.\n"); }
	printf("You can delete these rows with the following queries:\n");
	printf(query_delete_meta, "AGRID_META", "AGRID_ID", grid.grid_id);
	printf(query_delete_data, "AGRID", "AGRID_ID", grid.grid_id);
	
	printf("Inserting a bgrid...");
	if (nshm_put_bgrid(&grid)==NSHM_RETURN_SUCCESS) {
		printf("success!\n"); } else { printf("failed.\n"); }
	printf("You can delete these rows with the following queries:\n");
	printf(query_delete_meta, "BVALUE_META", "BVAL_ID", grid.grid_id);
	printf(query_delete_data, "BVALUE", "BVAL_ID", grid.grid_id);

	printf("Inserting an Mmax grid...");
	if (nshm_put_mmax(&grid)==NSHM_RETURN_SUCCESS) {
		printf("success!\n"); } else { printf("failed.\n"); }
	printf("You can delete these rows with the following queries:\n");
	printf(query_delete_meta, "MMAX_META", "MMAX_ID", grid.grid_id);
	printf(query_delete_data, "MMAX", "MMAX_ID", grid.grid_id);

	nshm_free_grid(&grid);
	nshm_cleanup();

	return EXIT_SUCCESS;
}

void make_small_grid(NSHM_Grid *_grid) {
	char * name = "Test Sample Grid";

	// This grid setup will have 12 points in it
	_grid->lat_min = 30.0;
	_grid->lat_max = 40.0;
	_grid->lat_inc = 2.0;
	_grid->lng_min = -120.0;
	_grid->lng_max = -118.0;
	_grid->lng_inc = 2.0;

	_grid->grid_values = calloc(12, sizeof(double));
	_grid->grid_values[0] = 1.0;
	_grid->grid_values[1] = 2.0;
	_grid->grid_values[2] = 3.0;
	_grid->grid_values[3] = 4.0;
	_grid->grid_values[4] = 5.0;
	_grid->grid_values[5] = 6.0;
	_grid->grid_values[6] = 7.0;
	_grid->grid_values[7] = 8.0;
	_grid->grid_values[8] = 9.0;
	_grid->grid_values[9] = 10.0;
	_grid->grid_values[10] = 11.0;
	_grid->grid_values[11] = 12.0;

	_grid->grid_name = calloc(strlen(name) + 1, sizeof(char));
	strncpy(_grid->grid_name, name, strlen(name));
}

File: nshm/src/tests/test_grid.f
      PROGRAM MAIN
      USE Grid
      IMPLICIT NONE

C --------------------------------------------------------------------
C --- Define some variables to be used throughout.                 ---
C --------------------------------------------------------------------

      INTEGER :: num_args
      CHARACTER*25 :: username, password, tnsalias
      CHARACTER*50 :: agrid_name, bgrid_name, mmax_name

      INTEGER :: num_rows, i

      TYPE(NSHM_Grid) :: grd
      num_args = IARGC()

C --- Check for proper usage
      IF (num_args.LT.3) THEN
         WRITE(*,100)
         STOP
      ENDIF

C --- Read in command line arguments
      CALL GETARG(1, username)
      username = TRIM(username)

      CALL GETARG(2, password)
      password = TRIM(password)

      CALL GETARG(3, tnsalias)
      tnsalias = TRIM(tnsalias)

C --- Set agrid, bgrid, and mmax grid names to fetch
      agrid_name = "adapt_cn"
      bgrid_name = "BVAL_.1"
      mmax_name  = "gm_j_6p6_7p1"

C --------------------------------------------------------------------
C --- Begin main body of program.                                  ---
C --------------------------------------------------------------------


C --- You must always initialize before using the API
      CALL nshm_initialize(username, password, tnsalias)

      WRITE(*,101)
      WRITE(*,102)"Fetching an Agrid"
      WRITE(*,101)
C --- Try to get an agrid
      CALL nshm_get_agrid(grd, agrid_name)
      CALL nshm_print_grid(grd)

C --- Try to write a copy of it back like it was a new grid
C      CALL nshm_put_agrid(grd)

      WRITE(*,101)
      WRITE(*,102)"Fetching a Bgrid"
      WRITE(*,101)
C --- Try to get a bgrid
      CALL nshm_get_bgrid(grd, bgrid_name)
      CALL nshm_print_grid(grd)

C --- Try to write a copy of it back like it was a new grid
C      CALL nshm_put_bgrid(grd)

      WRITE(*,101)
      WRITE(*,102)"Fetching a mmax"
      WRITE(*,101)
C --- Try to get a mmax
      CALL nshm_get_mmax(grd, mmax_name)
      CALL nshm_print_grid(grd)
 
C --- Try to write a copy of it back like it was a new grid
C      CALL nshm_put_mmax(grd)

C --- Check the number of rows returned from the last grid.
      CALL nshm_num_rows(grd, num_rows)
      WRITE(*,103)num_rows

C --- Natively in fortran echo the first 20 rows of the last grid
      WRITE(*,104)
      DO i=1, 20
         WRITE(*,105)grd%grid_values(i)
      END DO

C --- You should always cleanup when you are done with the API
      CALL nshm_cleanup()

C --------------------------------------------------------------------
C --- End main body of program. Below are format statements only.  ---
C --------------------------------------------------------------------

 100  FORMAT("Usage: test_grid_f <user> <pass> <tns>")
 101  FORMAT("######################################################")
 102  FORMAT("## ",A)
 103  FORMAT("The previous grid fetch returned ",I6," rows of data.")
 104  FORMAT("They are...")
 105  FORMAT(F10.7)

      END PROGRAM MAIN
